From 59c8f0412c45ad413f5cf4a5376a5e195162de25 Mon Sep 17 00:00:00 2001
From: Peter Antoine <peter.antoine@intel.com>
Date: Wed, 28 Oct 2015 14:04:28 -0700
Subject: [PATCH 1289/2508] MUST_REBASE [VPG]: drm/i915: MOCS table update and
 auto-generate support

commit 44253883ef32217450ccc568cc3ff02b27e1421c from
https://github.com/01org/linux-apollolake-i

The MOCS on Android must match the table within GmmLib. So this table will
need to match the version of GmmLib that it is intended to ship with. If
the tables do not match it will not break the system, just make it not as
preferment as it could be.

This change allows for the tables that are now generated by the UFO build
to be added easily with only updating a single standalone file. This file
is generated by the UFO build.

MUST_REBASE: This change is required for supporting android so that GmmLib
userspace based applications work as expected and is divergent from the
tables (and code) that is used in the upstream Linux kernel.

Change-Id: Ib0c9e9fb2c5c732bd7c9047e68606009925d5994
For: GMINL-13168
Signed-off-by: Peter Antoine <peter.antoine@intel.com>
[Jeff: Original patch included much of the base MOCS programming
implementation which is already present in this drm/i915 - that was
removed in the rebasing. Also excluded programming of the MOCS registers
in the GuC and WiDi spaces - revisit if/when necessary.]
Signed-off-by: Jeff McGee <jeff.mcgee@intel.com>
Signed-off-by: Matt Roper <matthew.d.roper@intel.com>
---
 drivers/gpu/drm/i915/Makefile     |    1 +
 drivers/gpu/drm/i915/intel_mocs.c |   44 -------------------------------------
 drivers/gpu/drm/i915/intel_mocs.h |   13 +++++++++++
 3 files changed, 14 insertions(+), 44 deletions(-)

diff --git a/drivers/gpu/drm/i915/Makefile b/drivers/gpu/drm/i915/Makefile
index 7ffb51b..ccb797f 100644
--- a/drivers/gpu/drm/i915/Makefile
+++ b/drivers/gpu/drm/i915/Makefile
@@ -37,6 +37,7 @@ i915-y += i915_cmd_parser.o \
 	  i915_trace_points.o \
 	  intel_lrc.o \
 	  intel_mocs.o \
+	  intel_mocs_gmm_table.o \
 	  intel_ringbuffer.o \
 	  intel_uncore.o
 
diff --git a/drivers/gpu/drm/i915/intel_mocs.c b/drivers/gpu/drm/i915/intel_mocs.c
index 7c7ac0a..e9a0547 100644
--- a/drivers/gpu/drm/i915/intel_mocs.c
+++ b/drivers/gpu/drm/i915/intel_mocs.c
@@ -24,17 +24,6 @@
 #include "intel_lrc.h"
 #include "intel_ringbuffer.h"
 
-/* structures required */
-struct drm_i915_mocs_entry {
-	u32 control_value;
-	u16 l3cc_value;
-};
-
-struct drm_i915_mocs_table {
-	u32 size;
-	const struct drm_i915_mocs_entry *table;
-};
-
 /* Defines for the tables (XXX_MOCS_0 - XXX_MOCS_63) */
 #define LE_CACHEABILITY(value)	((value) << 0)
 #define LE_TGT_CACHE(value)	((value) << 2)
@@ -126,39 +115,6 @@ static const struct drm_i915_mocs_entry broxton_mocs_table[] = {
 	  (L3_ESC(0) | L3_SCC(0) | L3_CACHEABILITY(L3_WB)) }
 };
 
-/**
- * get_mocs_settings()
- * @dev:        DRM device.
- * @table:      Output table that will be made to point at appropriate
- *              MOCS values for the device.
- *
- * This function will return the values of the MOCS table that needs to
- * be programmed for the platform. It will return the values that need
- * to be programmed and if they need to be programmed.
- *
- * Return: true if there are applicable MOCS settings for the device.
- */
-static bool get_mocs_settings(struct drm_device *dev,
-			      struct drm_i915_mocs_table *table)
-{
-	bool result = false;
-
-	if (IS_SKYLAKE(dev) || IS_KABYLAKE(dev)) {
-		table->size  = ARRAY_SIZE(skylake_mocs_table);
-		table->table = skylake_mocs_table;
-		result = true;
-	} else if (IS_BROXTON(dev)) {
-		table->size  = ARRAY_SIZE(broxton_mocs_table);
-		table->table = broxton_mocs_table;
-		result = true;
-	} else {
-		WARN_ONCE(INTEL_INFO(dev)->gen >= 9,
-			  "Platform that should have a MOCS table does not.\n");
-	}
-
-	return result;
-}
-
 static i915_reg_t mocs_register(enum intel_engine_id ring, int index)
 {
 	switch (ring) {
diff --git a/drivers/gpu/drm/i915/intel_mocs.h b/drivers/gpu/drm/i915/intel_mocs.h
index 76e45b1..c191656 100644
--- a/drivers/gpu/drm/i915/intel_mocs.h
+++ b/drivers/gpu/drm/i915/intel_mocs.h
@@ -52,6 +52,19 @@
 #include <drm/drmP.h>
 #include "i915_drv.h"
 
+struct drm_i915_mocs_entry {
+	u32	control_value;
+	u16	l3cc_value;
+};
+
+struct drm_i915_mocs_table {
+	u32					size;
+	const struct drm_i915_mocs_entry	*table;
+};
+
+bool get_mocs_settings(struct drm_device *dev,
+			struct drm_i915_mocs_table *table);
+
 int intel_rcs_context_init_mocs(struct drm_i915_gem_request *req);
 
 #endif
-- 
1.7.5.4

